localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -4118,87 +4141,141 @@ def delete_function_concurrency(
+                f"1 validation error detected: Value '{resource}' at 'resource' failed to satisfy constraint: Member must satisfy regular expression pattern: {api_utils.TAGGABLE_RESOURCE_ARN_PATTERN}"
             )
 
-        groups = pattern_match.groupdict()
-        fn_name = groups.get("function_name")
+        # Check whether the ARN we have been passed is correctly formatted
+        parsed_resource_arn: ArnData = None
+        try:
+            parsed_resource_arn = parse_arn(resource)
+        except Exception:
+            _raise_validation_exception()
+
+        # TODO: Should we be checking whether this is a full ARN?
+        region, account_id, resource_type = map(
+            parsed_resource_arn.get, ("region", "account", "resource")
+        )
 
-        if groups.get("qualifier"):
+        if not all((region, account_id, resource_type)):
+            _raise_validation_exception()
+
+        if not (parts := resource_type.split(":")):
+            _raise_validation_exception()
+
+        resource_type, resource_identifier, *qualifier = parts
+        if resource_type not in {"event-source-mapping", "code-signing-config", "function"}:
+            _raise_validation_exception()
+
+        if qualifier:
+            if resource_type == "function":
+                raise InvalidParameterValueException(
+                    "Tags on function aliases and versions are not supported. Please specify a function ARN.",
+                    Type="User",
+                )
+            _raise_validation_exception()
+
+        match resource_type:
+            case "event-source-mapping":
+                self._get_esm(resource_identifier, account_id, region)
+            case "code-signing-config":
+                raise NotImplementedError("Resource tagging on CSC not yet implemented.")
+            case "function":
+                self._get_function(
+                    function_name=resource_identifier, account_id=account_id, region=region
+                )
+
+        # If no exceptions are raised, assume ARN and referenced resource is valid for tag operations
+        return lambda_stores[account_id][region]
+
+    def tag_resource(
+        self, context: RequestContext, resource: TaggableResource, tags: Tags, **kwargs
+    ) -> None:
+        if not tags:
             raise InvalidParameterValueException(
-                "Tags on function aliases and versions are not supported. Please specify a function ARN.",
-                Type="User",
+                "An error occurred and the request cannot be processed.", Type="User"
             )
+        self._store_tags(resource, tags)
 
-        account_id, region = api_utils.get_account_and_region(resource, context)
-        fn = self._get_function(function_name=fn_name, account_id=account_id, region=region)
-
-        self._update_tags(fn, tags)
+        if (resource_id := extract_resource_from_arn(resource)) and resource_id.startswith(
```

Comment:
Unfortunately we need to run these dirty checks to ensure a Revision ID gets updated

Reply:
+1, good idea